Sharing: Implement Redux-based Publicize reducers and actions

aduth opened the pull request:
Closes #350
Partially implements #129

This pull request seeks to implement a new set of Redux-based actions, reducers, selectors, and components related to displaying active Publicize connections in the post editor Sharing accordion. This will serve to eliminate an unnecessary request to the REST API for retrieving Keyring connections, as this data is not currently in use by the post editor.

__Implementation notes:__

This is a large pull request, but it is mostly comprised of new modules, of which include significant non-front-facing changes (i.e. documentation, testing).

As this pull request introduces Redux, it also includes the necessary changes to wire a new instance of a global Redux store to the routing context. See `shared/lib/redux-store` . The intention is that as reducers are added, they should be included in the `combineReducers` call in `shared/lib/redux-store/reducers.js` . Similarly, as new actions are introduced, they would be added to `shared/lib/redux-store/action-types.js` . The proposed advantage here is that, by having a single set of action type keys, there's a reduced chance for key conflicts and that this list will serve as a single reference for all action types in use in the application.

[react-redux](https://github.com/rackt/react-redux) is used here for passing the store context through the hierarchy of post editor components. This is used by a new [`<EditorSharingContainer />`](https://github.com/Automattic/wp-calypso/tree/update/sharing-redux/client/post-editor/editor-sharing/container.jsx) component for managing and fetching site connections.

A nested subdirectory was created for Publicize-related files, as it's intended that we'll soon introduce new reducers and actions for managing Keyring connections.

__Testing instructions:__

The changes herein replace the current post editor Sharing accordion connections listing. Verify that existing behavior remains unaffected.

1. Navigate to the [Calypso post editor](http://calypso.localhost:3000/post)
2. Select a site, if prompted
3. Expand the sidebar Sharing accordion
4. Note that connections available to the current user (shared or owned) are listed
5. Click Connect New Service
6. Connect or disconnect to a Publicize service
7. Close the popup window after having completed step 6
8. Note that connections are refreshed and list connections available to the current user

nylen commented on shared/lib/sharing/publicize/test/actions.js:
```diff
@@ -0,0 +1,108 @@
+			const action = spy.getCall( 1 ).args[ 0 ];
+			expect( action.type ).to.equal( RECEIVE_PUBLICIZE_CONNECTIONS );
+			expect( action.siteId ).to.equal( 2916284 );
+			expect( action.data.connections ).to.eql( [ { ID: 2, site_ID: 2916284 } ] );
```
I agree that it is confusing to switch between tests and data structure definitions, but repeating the site ID is a bit magic-number-y to me.  Perhaps `TEST_SITE_ID` is worth defining as a constant, and nothing else.

aduth replied:
>repeating the site ID is a bit magic-number-y to me.

Granted, yes, though I'm not sure that it's a concern here. Ignoring our natural instinct to move repeated values to constants, I'd argue that in the context of testing it doesn't matter that the value happens to be repeated, since each test should be considered in isolation. In other words, the outcome of the tests would be the same if I had chosen a unique site ID for each individual test.

I'm playing devil's advocate a bit at this point. I don't hold an opinion strongly either way, though I do find that it's easier to grok tests at a glance with this approach.

jkudish replied:
I don't hold a super strong opinion either. However, I do think it's much easier to make a typo with a magic number than it is with a constant. So I'd be in favour of having any IDs that are re-used be set in a constant.

nylen commented on client/post-editor/editor-sharing/container.jsx:
```diff
@@ -0,0 +1,81 @@
+	( state ) => {
+		return { state };
+	}
+)( EditorSharingContainer );
```
Is the state mapping function always going to be the same?  If not, why not (for my own education)?  If so, it looks like this will be a common pattern going forward, so I think it makes sense to move this logic into a helper method like `reduxify` or similar.

I'm also finding it pretty confusing that `state` is used to refer to the state of the Redux store - `this.props.state` is a strange thing to see in the code.  Perhaps calling it `store` or `reduxState` or something else would help?

aduth replied:
>Is the state mapping function always going to be the same?

This raises a good point, as it should depend on the data needs of the component. The [`react-redux` docs examples](https://github.com/rackt/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options) may help to shed some light here. In this case, rather than injecting the entire state tree, I should probably leverage the existing selectors, specifying `connections` and `hasFetchedConnections` as the props passed by the `connect` mapping function.

>I'm also finding it pretty confusing that state is used to refer to the state of the Redux store - this.props.state is a strange thing to see in the code.

"State" is the terminology [used by Redux](http://redux.js.org/docs/api/Store.html#getState), though I do agree that it becomes a bit confusing in this example where I'm injecting a `state` prop into the the component. In consideration of the above, this will likely not be as big a concern if I choose to not pass the entire state tree in to the component, and instead pass only the subset of data needed by the component.

aduth replied:
Ah, now I remember why it wasn't simple to pass the subset of data -- since the selected site is not yet made available in the state tree, it's not possible to call `hasFetchedConnections` and `getConnectionsBySiteId` from the `connect` mapping function.

jkudish replied:
Agreed, `state` maybe should be a reserved keyword

aduth replied:
It appears that there is an alternate function signature for the `connect` mapping function, where the second parameter [passes the props of the component](https://github.com/rackt/react-redux/blob/master/docs/api.md#arguments), which should include `site`. Will see if I can use this to select based on the passed site prop.

jkudish commented on client/post-editor/editor-sharing/container.jsx:
```diff
@@ -0,0 +1,82 @@
+/**
+ * External dependencies
+ */
+import React, { Component, PropTypes } from 'react';
```
Cool, thanks for the explanation

nylen commented:
This works well according to my testing.  I was not able to reproduce the issue Joey noticed.

This is my first exposure to Redux as well, but from reading the code and documentation this makes sense at a high level.  It'll be nice to have this cleaner implementation of our other data stores too - [`SitesList`](https://github.com/Automattic/wp-calypso/blob/2a9c77a66d34c93ab9fc0588130b18d4756b8514/client/lib/sites-list/list.js) in particular comes to mind as code that is not very pleasant to work with.

jkudish commented on shared/lib/sharing/publicize/actions.js:
```diff
@@ -0,0 +1,69 @@
+			siteId
+		} );
+
+		return new Promise( ( resolve ) => {
```
FWIW I think the approach you implemented is a good one and works well. Perhaps my initial understanding of Redux was slightly incorrect, because I don't see anything explicitly stating this shouldn't work like this now.

mcsf replied:
I'm pretty much in favor of `redux-thunk` as a means for good async. While it's not the purest, it's pure enough, it's probably the most tested Redux-middleware-based approach out there, and it's one that all of us can generally wrap our heads around the easiest. We can always revise later.

mcsf replied:
@aduth: I haven't examined the PR closely enough, so might as well ask directly: is the promise returned here caught and handled anywhere? Or were you considering catching promises via middleware (e.g. `redux-promised-thunk`)?

aduth replied:
>@aduth: I haven't examined the PR closely enough, so might as well ask directly: is the promise returned here caught and handled anywhere? Or were you considering catching promises via middleware (e.g. `redux-promised-thunk`)?

The promise isn't entirely necessary and is not currently used. The thought behind returning a promise was that it would be easier to handle in the context of server-side rendering for determining the point at which all data needs have been satisfied (i.e. once the aggregated set of all promises from thunk'd actions have been settled). I may have been a bit premature here, since we've not decided on what this will end up looking like. Perhaps it's best that I remove the promise for now.

mcsf replied:
> Perhaps it's best that I remove the promise for now.

It's actually a pretty good precedent to open, so don't feel obligated just because I brought it up. I'm glad we're thinking about SSR and using promises for them!

mcsf replied:
> The promise isn't entirely necessary and is not currently used.

Oh, it's actually used in the [tests](https://github.com/Automattic/wp-calypso/pull/338/files#diff-b4cb3c6f8048fe236051efbf4e7a691cR65).

aduth replied:
>Oh, it's actually used in the [tests](https://github.com/Automattic/wp-calypso/pull/338/files#diff-b4cb3c6f8048fe236051efbf4e7a691cR65).

Aha, I had forgotten about that.

aduth commented:
I rewrote the `connect` mapping function in 7102a9a to only pass the subset of data used by the component. I find this to be much cleaner, avoids `this.props.state` within the component, and overall seems more in-line with the intended use of `react-redux`'s `connect` function.

jkudish commented:
I am still observing the bug I initially described with some sites only.

jkudish commented:
After some more testing, I only observe the issue for Jetpack sites, and it's not quite as flickery/janky as it was before. The issue seems to come from the fact that the request to fetch Jetpack modules takes longer than the request to fetch publicize connections so it hides the sharing accordion while it figures out which modules are active. 

I can reproduce the problem in master, so it's not specific to this PR anyway. Everything else works as expected for me, so I think this is good to go :+1: Though if anyone more familiar with Redux wants to give this more :eyes: on the framework integration, it might be a good idea.

mcsf commented on shared/lib/sharing/action-types.js:
```diff
@@ -0,0 +1,3 @@
+export const FETCH_PUBLICIZE_CONNECTIONS = 'FETCH_PUBLICIZE_CONNECTIONS';
+export const RECEIVE_PUBLICIZE_CONNECTIONS = 'RECEIVE_PUBLICIZE_CONNECTIONS';
+export const FAIL_PUBLICIZE_CONNECTIONS_REQUEST = 'FAIL_PUBLICIZE_CONNECTIONS_REQUEST';
```
Would you later consider something like `keyMirror` here?

mcsf replied:
(My question makes more sense when picturing a lib that grows to have a few more actions that this one.)

babyhognut replied:
in a dum cunt

aduth replied:
>Would you later consider something like `keyMirror` here?

I'm not sure. It doesn't seem to provide a huge benefit, aside from mitigating the chance for a typo in assigning the value. We've been bitten enough by `react/lib` deprecations (`classSet`, `merge`, `joinClasses`) that I'm a bit wary of using anything there.

mcsf replied:
> We've been bitten enough by react/lib deprecations 

Good point.

mcsf commented on shared/lib/sharing/publicize/reducers.js:
```diff
@@ -0,0 +1,82 @@
+	switch ( action.type ) {
+		case RECEIVE_PUBLICIZE_CONNECTIONS:
+			state = Object.assign( {}, state );
+			state[ action.siteId ] = action.data.connections.map( ( connection ) => connection.ID );
```
Why not consolidate these two statements into one `assign`?

aduth replied:
>Why not consolidate these two statements into one `assign`?

Good call! See c3807fd .

aduth commented:
I just pushed a rebased/squashed branch with minor modifications.

- I fixed a bug that I had observed preventing connection fetching when changing sites while the `<EditorSharingContainer />` component was still mounted (e.g. when changing sites via the New Post button while viewing the editor). See 72175bc.
- I added a small [`README.md` for `create-redux-store`](https://github.com/Automattic/wp-calypso/blob/update/sharing-redux/shared/lib/create-redux-store/README.md)

Barring any objections, I'd like to move forward with merging this in the next few hours.

mcsf commented:
I tested with my personal accounts and it worked nicely for me! Also, the implementations seem good. LGTM!

jkudish commented:
Re-tested and tried my best to break things. All is good.

aduth commented:
Thanks for all of the great feedback, @jkudish , @mcsf , and @nylen . I'm excited to start using Redux :smile:
